[MRG] Matern kernel

jmetzen opened the pull request:
Add Matern kernel to the kernels in pairwise.py.

The Matern kernel (https://en.wikipedia.org/wiki/Mat%C3%A9rn_covariance_function and Rasmussen and Williams 2006, pp84) is a generalization of the RBF and the absolute exponential kernel with an additional hyperparemeter nu, which allows interpolating between these two (RBF: nu=inf, absolute exponential: nu=0.5). In contrast to the RBF kernel, it makes less strict assumptions on the smoothness of the function to be learned. This is shown in an example for a step-function for different values of nu:

![figure_0](https://cloud.githubusercontent.com/assets/1116263/5189587/64c46cb2-74df-11e4-9b94-684e9751bc78.png)

TODOs:
- [x] support for arbitrary values of nu (coef0)
- [x] tests

Commit:
ADD Matern kernel added to pairwise.py

REFACTOR Performance improvements in matern_kernel based on @afabisch's suggestions

EXAMPLE Comparing Matern kernels of different smoothness on a step-function.

ENH Support for arbitrary values of coef0 ("nu") in the Matern kernel

EXAMPLE Added illustration of Matern kernel for different values of coef0

Commit:
DOC Adding documentation for Matern kernel

DOC Extended documentation of Matern kernel

DOC Documentation of parameter coef0 of matern_kernel

Commit:
TEST Adding tests for Matern kernel

MISC Using np.finfo(float).eps instead of 1e-10 in Matern kernel

Furthmore, fixed typo and increased readability of test_matern_kernel.

TEST Extended tests of Matern kernel

amueller commented on sklearn/metrics/tests/test_pairwise.py:
```diff
@@ -415,6 +416,22 @@ def test_rbf_kernel():
     assert_array_almost_equal(K.flat[::6], np.ones(5))
 
 
+def test_matern_kernel():
```
maybe add an absolute exponential kernel test?

amueller replied:
Actually at least all if branches need testing.

jmetzen replied:
All if branches are now tested (line 437ff). The absolute exponential test (line 430ff) is a bit self-referential; an alternative would be to test against the absolute_exponential in gaussian_process.correlation_models.

amueller commented on examples/metrics/plot_matern_kernel.py:
```diff
@@ -0,0 +1,40 @@
+plt.yscale("log")
+plt.ylim(1e-3, 1e0)
+plt.legend(title="coef0")
+plt.show()
```
no newline

agramfort replied:
+1

amueller commented:
Maybe a stupid question, but what it this meant to be used for? SVMs? Is that practical?
I guess this is a pretty standard kernel so I'm not against adding it, I'm just not entirely certain of the usefulness in the current form.

Commit:
PEP8 conformity improved

DOC Add "matern" as valid kernel name in various places in the doc

MISC Revising Matern kernel PR based on @agramfort's review

REFACTOR Using K.fill(X) instead of K[:] = X

jmetzen commented:
I believe the main use-case of the Matern kernel is in combination with Gaussian processes. Combined with GPs, the Matern kernel (nu=1.5 or 2.5) can often improve considerably over the RBF-kernel and absolute exponential kernel (see example in PR #3388), since it has weaker assumptions on the smoothness of the function than RBF and is less jumpy than the absolute exponential kernel. That being said, it can also be beneficial with other kernel-methods. I've used it together with kernel_approximation.Nystroem where it also often outperformed the RBF-kernel.

Once this PR is merged, I would like to work on refactoring correlation_models.py such that it uses as much from pairwise.py as possible (in particular the Matern kernel). I'm not doing that directly in this PR in order to keep this PR small and focused.

amueller commented:
Ok, fair enough. It does seem to me that the matern kernel is mostly used in GPs.

AlexanderFabisch commented:
Although it is usually used in GPs, I think we should integrate it in the library. It seems to be useful in combination with e.g. kernel ridge regression and seems to be a basic kernel even though it might not be used as often as the RBF kernel. I would +1 for merge.

jmetzen commented:
Yes, would be nice if we could merge this PR.  @amueller What do you think?

amueller commented:
Could you squash and rebase please?

amueller commented:
And yeah, we can merge it.

amueller commented:
Have you checked that the docs are rendered properly?

amueller commented:
We want to release pretty soon so we are focusing on bug-fixes a bit currently, and I'm a bit busy the next two weeks. I'll try to review but I can't promise anything, sorry. After the release, I'll certainly have the time.

jmetzen commented:
Done. The documentation is rendered properly for me.

agramfort commented on examples/metrics/plot_matern_kernel.py:
```diff
@@ -0,0 +1,40 @@
+
+from sklearn.metrics.pairwise import matern_kernel
+
+import matplotlib.pyplot as plt
```
matplotlib import should be before sklearn

agramfort commented on examples/svm/plot_svm_matern_kernel.py:
```diff
@@ -0,0 +1,73 @@
+from sklearn.svm import NuSVR
+from sklearn.metrics.pairwise import matern_kernel
+
+import matplotlib.pyplot as plt
```
same here

agramfort commented on sklearn/metrics/pairwise.py:
```diff
@@ -767,6 +769,72 @@ def rbf_kernel(X, Y=None, gamma=None):
+
+    gamma : float
+
+    coef0 : float>0.0 (the parameter nu)
```
this formatting is not standard

AlexanderFabisch replied:
All parameter descriptions in this file seem to ignore the standard. Do you think we should change that in this or another pull request?

agramfort replied:
if you don't do it now, create a PR asap after we merge or create an issue
so we don't forget.

agramfort commented on sklearn/metrics/pairwise.py:
```diff
@@ -767,6 +769,72 @@ def rbf_kernel(X, Y=None, gamma=None):
+
+    Parameters
+    ----------
+    X : array of shape (n_samples_X, n_features)
```
Formatting should be

X : array, shape (n_samples_X, n_features)

agramfort commented on sklearn/metrics/pairwise.py:
```diff
@@ -767,6 +769,72 @@ def rbf_kernel(X, Y=None, gamma=None):
+    ----------
+    X : array of shape (n_samples_X, n_features)
+
+    Y : array of shape (n_samples_Y, n_features)
```
same here and below

agramfort commented on sklearn/metrics/pairwise.py:
```diff
@@ -767,6 +769,72 @@ def rbf_kernel(X, Y=None, gamma=None):
+        K *= -gamma
+        np.exp(K, K)  # exponentiate K in-place
+    elif coef0 == 1.5:
+        K *= np.sqrt(3) * gamma
```
for scalars use sqrt from math module

agramfort commented on sklearn/metrics/pairwise.py:
```diff
@@ -767,6 +769,72 @@ def rbf_kernel(X, Y=None, gamma=None):
+        np.exp(K, K)  # exponentiate K in-place
+    elif coef0 == 1.5:
+        K *= np.sqrt(3) * gamma
+        K = (1 + K) * np.exp(-K)
```
1 -> 1.

agramfort commented on sklearn/metrics/pairwise.py:
```diff
@@ -767,6 +769,72 @@ def rbf_kernel(X, Y=None, gamma=None):
+        K *= np.sqrt(3) * gamma
+        K = (1 + K) * np.exp(-K)
+    elif coef0 == 2.5:
+        K *= np.sqrt(5) * gamma
```
math.sqrt

agramfort commented:
that's it for me.

jmetzen commented:
Thanks for the review @agramfort . Your comments should be addressed. I changed the format to "X: array, shape (n_samples_X, n_features)" etc. in the whole pairwise module.

agramfort commented on sklearn/metrics/pairwise.py:
```diff
@@ -767,6 +768,72 @@ def rbf_kernel(X, Y=None, gamma=None):
+    else:  # general case; expensive to evaluate
+        K[K == 0.0] += np.finfo(float).eps  # strict zeros would result in nan
+        tmp = (math.sqrt(2 * coef0) * gamma * K)
+        K[:] = (2 ** (1. - coef0)) / scipy.special.gamma(coef0)
```
I would have used K.fill 

besides LGTM

jmetzen replied:
done

amueller commented:
was the intention here mostly to use it with the GP and this is replaced by #4270 or do we still want this in pairwise?

glemaitre commented:
Since we already have the Matern kernel in GPR and I did not find any issue requesting the kernel for another usage, I am closing this PR.
